### PR TITLE
feat[next]: use context vars instead of global state in embedded iterator execution

### PR DESCRIPTION
**Pull request by havogt**
Changed:
- In the iterator embedded execution, both `column_range` and `offset_provider` global variables have been converted to context variables and fencil closures are now executed in isolated contexts with properly initialized values, to avoid using global state.

### CODE REVIEW

**Comment by havogt** [comment]
Opening as a reminder, instead of #896

**Comment by egparedes** [comment]
I can clean up the code and provide a final version of this prototype without too much effort. @tehrengruber  do you also agree with using this implementation strategy to fix the problem?

**Comment by tehrengruber** [comment]
Yes, given the model itself this is much better than the global variable.

**Comment by egparedes** [comment]
I've requested a formal review from @tehrengruber because apparently @havogt can't approve it since he's the creator of the PR, but I think both of you should take a look.

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:187]: I‘m not convinced of the naming scheme, but your choice.

**Review by tehrengruber — APPROVED** [review]

File: src/gt4py/next/iterator/embedded.py
Comment (egparedes) [reply src/gt4py/next/iterator/embedded.py:187]: Any suggestion?

Comment (egparedes) [reply src/gt4py/next/iterator/embedded.py]: Changed

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:202]:
```suggestion
        column_range = column_range_cvar.get()
```
Since it is a local variable now the underscore is not needed anymore right?

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:726]:
```suggestion
    column_range = column_range_cvar.get()
```
Same as above.

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:775]:
```suggestion
        offset_provider = offset_provider_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:800]:
```suggestion
        column_range = column_range_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:843]:
```suggestion
        column_range = column_range_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:1093]:
```suggestion
    offset_provider = offset_provider_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:1152]:
```suggestion
        offset_provider = offset_provider_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:1302]:
```suggestion
        column_range = column_range_cvar.get()
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:1348]:
```suggestion
        column_range = None
```

File: src/gt4py/next/iterator/embedded.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/embedded.py:1334]:
```suggestion
    offset_provider = kwargs["offset_provider"]
```

File: tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py
Comment (tehrengruber) [inline tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py:153]: Is this really a meaningful test? I have no clue how the thread scheduling works in Python, but this might very well never fail even if the column range is implemented incorrectly. If we want this to be meaningful how about a lock that ensure the column construction of the first thread happens after the context setup of a second thread? (If it's not worth the effort I would remove the test altogether to be honest.)

File: tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py
Comment (tehrengruber) [inline tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py:49]: This is fine, but is there a reason to not use a context handler for this? Sure the columns would leak scope, but doesn't really matter right?

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py
Comment (egparedes) [reply tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py:153]: Ok, I agree and removed the test.

File: tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py
Comment (egparedes) [reply tests/next_tests/unit_tests/iterator_tests/test_embedded_internals.py:49]: A context handler doesn't really work here because code running inside a context needs to be an actual function (line 36). Setting the values of the context vars before running some lines of code and then restoring their previous values does not test the context vars and it doesn't work in the general case (it will fail when running multiple threads, for example).

**Review by tehrengruber — APPROVED** [review]

**Pull request merged**
